provider/apple.go

```diff
@@ -0,0 +1,534 @@
+	Params
+
+	// all of these fields specific to particular oauth2 provider
+	name string
+	// infoURL  string not implemented at Apple side
+	endpoint oauth2.Endpoint
+
+	mapUser func(jwt.MapClaims) token.User // map info from InfoURL to User
+	conf    AppleConfig                    // main config for Apple auth provider
+
+	PrivateKeyLoader PrivateKeyLoaderInterface // custom function interface for load private key
+
+}
+
+// PrivateKeyLoaderInterface interface for implement custom loader for Apple private key from user source
+type PrivateKeyLoaderInterface interface {
+	LoadPrivateKey() ([]byte, error)
+}
+
+// LoadFromFileFunc is the type for use pre-defined private key loader function
+// Path field must be set with actual path to private key file
+type LoadFromFileFunc struct {
+	Path string
+}
+
+// LoadApplePrivateKeyFromFile return instance for pre-defined loader function from local file
+func LoadApplePrivateKeyFromFile(path string) LoadFromFileFunc {
+	return LoadFromFileFunc{
+		Path: path,
+	}
+}
+
+// LoadPrivateKey implement pre-defined (built-in) PrivateKeyLoaderInterface interface method for load private key from local file
+func (lf LoadFromFileFunc) LoadPrivateKey() ([]byte, error) {
+	if lf.Path == "" {
+		return nil, errors.New("empty private key path not allowed")
+	}
+
+	kFile, err := os.Open(lf.Path)
+	if err != nil {
+		return nil, err
+	}
+	keyValue, err := ioutil.ReadAll(kFile)
+	if err != nil {
+		return nil, err
+	}
+	err = kFile.Close()
+	return keyValue, err
+}
+
+// NewApple create new AppleProvider instance with custom name and parameters
+// endpoints used in constructor  required only for testing and can't be define one outside the package.
+// Private key must be set, when instance create call, for create `client_secret`
+func NewApple(name string, p Params, appleCfg AppleConfig, endpoints oauth2.Endpoint, privateKeyLoader PrivateKeyLoaderInterface) (*AppleHandler, error) {
+
+	if name == "" {
+		return nil, errors.New("empty name for apple provider not allowed")
+	}
+
+	if p.L == nil {
+		p.L = logger.NoOp
+	}
+
+	// configuring default values
+	if appleCfg.UserAgent == "" {
+		appleCfg.UserAgent = defaultUserAgent
+	}
+	if endpoints.AuthURL == "" {
+		endpoints.AuthURL = appleAuthURL
+	}
+	if endpoints.TokenURL == "" {
+		endpoints.TokenURL = appleTokenURL
+	}
+
+	ah := AppleHandler{
+		Params: p,
+		name:   name,
+
+		conf: AppleConfig{
+			ClientID:           appleCfg.ClientID,
+			TeamID:             appleCfg.TeamID,
+			KeyID:              appleCfg.KeyID,
+			ClientSecretExpire: appleCfg.ClientSecretExpire,
+			UserAgent:          appleCfg.UserAgent,
+			Scopes:             appleCfg.Scopes,
+		},
+
+		endpoint: endpoints,
+		mapUser: func(claims jwt.MapClaims) token.User {
+			var usr token.User
+			if uid, ok := claims["sub"]; ok {
+				usr.ID = fmt.Sprintf("apple_%s", uid.(string))
+			}
+
+			// this claim exist if "email" defined in scope
+			if email, ok := claims["email"]; ok {
```

Comment:
this one is for what exactly? Why we even need it and how email_verified (for verified account with the manual confirmation) even related to this?

Reply:
it's claim return in token with other claims. I add it as additional flag for user check. If it isn't need i remove it.

Reply:
well, in fact, it exposes email in the final token. And I don't get the part about email_verified, is it smth apple provides?

Reply:
Yes. Apple return token with this additional claim. Email claim exist as well.
[Token description](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_rest_api/authenticating_users_with_sign_in_with_apple#3383773)